tests/ui/unnecessary_map_on_constructor.rs

```diff
@@ -0,0 +1,43 @@
+#![allow(unused)]
+#![warn(clippy::unnecessary_map_on_constructor)]
+
+fn fun(t: i32) -> i32 {
+    t
+}
+
+fn notfun(e: SimpleError) -> SimpleError {
+    e
+}
+
+#[derive(Copy, Clone)]
+struct SimpleError {}
+
+type SimpleResult = std::result::Result<i32, SimpleError>;
+
+fn main() {
+    let x: i32 = 4;
+    //
+    let err = SimpleError {};
+    let a = Some(x); // => Some(fun(x))
+    let b: SimpleResult = Ok(x); // => Ok(fun(x))
+    let c: SimpleResult = Err(err); // => Ok(fun(x))
+    //
+    let a = Some(x).map(fun); // => Some(fun(x))
+    let b: SimpleResult = Ok(x).map(fun); // => Ok(fun(x))
+    let c: SimpleResult = Err(err).map_err(notfun); // => Ok(fun(x))
+    //
+    //
+    let a = Option::Some(x).map(fun); // => Some(fun(x))
+    let b: SimpleResult = SimpleResult::Ok(x).map(fun); // => SimpleResult::Ok(fun(x))
+    let c: SimpleResult = SimpleResult::Err(err).map_err(notfun); // => SimpleResult::Err(notfun(x))
+    let b: std::result::Result<i32, SimpleError> = Ok(x).map(fun); // => Ok(fun(x))
+    let c: std::result::Result<i32, SimpleError> = Err(err).map_err(notfun); // => Err(notfun(x))
+
+    a.map(fun); // these should not trigger
+    b.map(fun);
+    c.map_err(notfun);
+
+    let a = Some(fun(x));
+    let b: SimpleResult = Ok(fun(x));
+    let c: SimpleResult = Err(notfun(err)); // => Ok(fun(x))
```

Comment:
The comments in the test can be dropped, some of them are incorrect and we'll be adding a standard way to denote it [in the near future](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Improve.20UI.20error.20checks)

Reply:
[d4c2a9d](https://github.com/rust-lang/rust-clippy/pull/11413/commits/d4c2a9db436caa026f7d54123633da1da760ebcf)